Set SELinux types on cri-o files

cevich opened the pull request:
After building/installing CRI-O, set the SELinux types on
specific files based on platform (ansible_distribution) name.
If no mapping for that platform is available, choose the default
item if one is present.  Failing both, skip the file.

For now this only includes /usr/local/bin/crio and sets the same
type on all platforms.  However this is easily expanded by
updating the mapping in ``vars.yml`` to include additional
files and/or ansible_distribution names (or "default") and types.

Signed-off-by: Chris Evich <cevich@redhat.com>

cevich commented:
@rhatdan your wish is my command (just make sure you wish really really loudly) 😀

cevich commented:
/test all

rhatdan commented:
LGTM

mrunalp commented:
We need to take out `setenforce 0` from e2e.yml to see if this really works.

rhatdan commented:
Agreed, and then capture the AVC messages generate during the run.

rhatdan commented:
ausearch -m avc, user_avc

mrunalp commented:
We already capture AVCs: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/kubernetes-incubator_cri-o/906/test_pull_request_crio_e2e_rhel/134/artifacts/generated/avc_denials.log

rhatdan commented:
Good.

rhatdan commented:
@cevich Looks like it is complaining about missing policycoreutils?

cevich commented:
Hmm, maybe I spelled it wrong.  Will take a look next week (and kill the ``setenforce 0``)

rhatdan commented:
I think you will need policycoreutils-python3 for Fedora.

cevich commented:
> I think you will need policycoreutils-python3 for Fedora.

Hmm, can try that, but mostly Ansible tries to stick to python2 stuff.  Looking at it now.

cevich commented:
Ahh, right, durr.  This is because of the setup-caching thing.  @stevekuznetsov I've added a package to the install list, but that's not going to be reflected in the cached AMI being used to test the PR.  Is there a way I can make the CI job run with all the tags, so it includes the ``setup`` tag?

Plan B: I'll stick in a dummy commit to just always install that package, for testing purposes.

cevich commented:
/test all

cevich commented:
Nevermind @stevekuznetsov @mrunalp answered my question.

stevekuznetsov commented on contrib/test/integration/build/cri-o.yml:
```diff
@@ -77,3 +83,17 @@
+    name: "libselinux-python, policycoreutils-python"
+    state: latest
+
+- name: Enforce specific SELinux types for crio files on this platform
```
Shouldn't this be handled by an RPM install?

cevich replied:
Unless I'm missing something, this is for testing from the compiled cri-o code.  So we need it both ways (in the RPM and in the playbook).  Though you're correct, if we're going to run tests after installing the RPM, this step isn't necessary.

stevekuznetsov replied:
Ah, I understand

cevich commented:
rebase ``--autosquash`` and ``--force`` push after killing the task: "Disable SELinux" 😀

cevich commented:
/test ami

cevich commented:
> Ah, I understand

Thanks for checking @stevekuznetsov , I know the other guys appreciate it, even if they don't say so all the time.

cevich commented:
hrmmm, the slash-test AMI console logs are wrong.  It shows this task running:

``TASK [disable SELinux] *********************************************************``

However, I deleted that task in 9a217dc ?!?!?!  Anyhoo, the good news is the new task I added appears to be working perfectly:

```
TASK [Enforce specific SELinux types for crio files on this platform] **********
task path: /go/src/github.com/kubernetes-incubator/cri-o/contrib/test/integration/build/cri-o.yml:87
changed: [localhost] => (item={'key': u'/usr/local/bin/crio', 'value': {u'default': u'container_runtime_exec_t'}}) => {
    "changed": true, 
    "ftype": "a", 
    "item": {
        "key": "/usr/local/bin/crio", 
        "value": {
            "default": "container_runtime_exec_t"
        }
    }, 
    "serange": "s0", 
    "setype": "container_runtime_exec_t", 
    "seuser": "system_u", 
    "state": "present", 
    "target": "/usr/local/bin/crio"
}
```

The bad news is, now the cri-o e2e tests fail 😖   

```


TASK [run e2e tests] ***********************************************************
task path: /go/src/github.com/kubernetes-incubator/cri-o/contrib/test/integration/e2e.yml:51
fatal: [localhost]: FAILED! => {
    "changed": true, 
    "cmd": "/usr/bin/go run hack/e2e.go --test -test_args=\"-host=https://172.18.4.21:6443 --ginkgo.focus=\\[Conformance\\] --report-dir=/tmp/artifacts\" &> /tmp/artifacts/e2e.log ", 
    "delta": "0:15:38.677798", 
    "end": "2017-10-09 19:51:37.367788", 
    "failed": true, 
    "rc": 1, 
    "start": "2017-10-09 19:35:58.689990"
}
	to retry, use: --limit @/go/src/github.com/kubernetes-incubator/cri-o/contrib/test/integration/main.retry

PLAY RECAP *********************************************************************
localhost                  : ok=27   changed=17   unreachable=0    failed=1 
```

I can't tell from that why it fails.  Maybe the ``e2e.log`` says something?

cevich commented:
No love, the ``e2e.log`` just says it got an error 😢 

```
2017/10/09 19:51:37 main.go:260: Something went wrong: encountered 1 errors: [error during ./hack/ginkgo-e2e.sh -host=https://172.18.4.21:6443 --ginkgo.focus=\[Conformance\] --report-dir=/tmp/artifacts: exit status 1]
2017/10/09 19:51:37 e2e.go:78: err: exit status 1
exit status 1
```

cevich commented:
More looking.  The ``ci/openshift-jenkins/e2e_rhel`` job isn't running the code from this PR:

```
+ git checkout master
Already on 'master'
+ git branch -D target
error: branch 'target' not found.
+ true
+ git branch -D pr
error: branch 'pr' not found.
+ true
+ git branch target 825f7f290b3c5518c2a55af25dad94b7f5f3efe9
+ git checkout target
Switched to branch 'target'
+ git fetch origin pull/982/head:pr
From https://github.com/kubernetes-incubator/cri-o
 * [new ref]         refs/pull/982/head -> pr
+ git merge 23086a19251e0244be5d8d7ea9f23c47ed71ad40
Merge made by the 'recursive' strategy.
```

So that's why that failed.

The other two (ami_rhel & ami_fedora) appear to be actual e2e-test failures, with different causes.  The collected ``generated/avc_denials.log`` for both is empty, so either no denials happened/logged or the job failed to collect them (somehow).  I don't see any full disks or weird things in ``pid1.journal``.  The ``crio.service`` log appears more interesting, with this for RHEL:

```
Oct 09 19:47:16 ip-172-18-4-21.ec2.internal crio[4910]: time="2017-10-09 19:47:16.240599991Z" level=error msg="Container creation error: container_linux.go:288: creating new parent process caused "container_linux.go:1655: running lstat on namespace path \"/proc/0/ns/ipc\" caused \"lstat /proc/0/ns/ipc: no such file or directory\""
```

and fedora:

```Oct 09 20:34:28 ip-172-18-9-162.ec2.internal crio[19102]: time="2017-10-09 20:34:28.536031672Z" level=error msg="Error deleting network: failed to Statfs "/var/run/netns/k8s_pod-service-account-mountsa_e2e-tests-svcaccounts-b035j_3bb025f5-ad31-11e7-b918-0ec5f7e221ee_0-3054fff3": no such file or directory"
```

No idea if those are real problems or not.  @rhatdan or @mrunalp could you or someone else take a look here please?    As near as I can tell the selinux-part of the playbook is working as requested, though perhaps more is needed (mass filesystem relabel?).

In any case, I'm basically blocked on this PR, so moving on to one of the others 'till someone tells me what's needed here 😃

rhatdan commented:
I don't think the issues you are seeing are necessarily SELinux related.  If you get permission denied, that is probably an issue.  But file not found is usually not SELinux related.

cevich commented:
Hmmm, so @rhatdan do you think some other change I made here could be causing the failures?

(I'll go double-check again...)

cevich commented:
I love it when running the exact same thing again, gives different results 😖   This time the RHEL e2e tests ran (two failures) but the Fedora integration tests timed out (didn't look at why).

cevich commented:
(rebase from master & ``--force`` push)

cevich commented:
/test ami

cevich commented:
(rebase from master & ``--force`` push)

cevich commented:
/test ami

cevich commented:
(ami_rhel result is bogus: provisioning failure 😖)

cevich commented:
/test ami

cevich commented:
/test ami(rebase from master & ``--force`` push)

cevich commented:
/test ami

cevich commented:
rebase from master & ``--force`` push

/test ami

cevich commented:
rebase from master & --force push

/test ami

cevich commented:
ami_rhel failed: provisioning problem

cevich commented:
/test ami_rhel

mrunalp commented:
/test ami

mrunalp commented:
@rhatdan I have made the changes to preseve the instances from the ami run for you to look at.

rhatdan commented:
@mrunalp Can you send me the info on how I can reach the machine?

cevich commented:
@rhatdan  When u get a moment:   I just looked at the RHEL results (assuming fedora ones are similar).  The integration tests are passing 100% but the e2e tests die with this result:

```/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/e2e.go:221
Oct 19 21:45:15.480: Error waiting for all pods to be running and ready: 1 / 1 pods in namespace "kube-system" are NOT in RUNNING and READY state in 10m0s
POD                       NODE      PHASE   GRACE CONDITIONS
kube-dns-3097350089-16jkx 127.0.0.1 Pending       [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2017-10-19 21:29:52 +0000 UTC  } {Ready False 0001-01-01 00:00:00 +0000 UTC 2017-10-19 21:29:52 +0000 UTC ContainersNotReady containers with unready status: [kubedns dnsmasq sidecar]} {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2017-10-19 21:29:52 +0000 UTC  }]

/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/e2e.go:172
```

I checked and there are no AVC denials (or, at least, none are reported).  Is this the same symptom y'all were discussing yesterday on IRC?

rhatdan commented:
Yes.  If you run a test with SELinux in enforcing mode.  I need to get onto the node to try to figure out what is going on.

cevich commented:
See also: #1064

cevich commented:
/test ami

cevich commented:
Closing in favor of #1064
